### ISSUE TITLE
Use `ykman` instead of `yubikey-personalization`

### ISSUE THREAD

**Issue by EliasHolzmann** [agherzan/yubikey-full-disk-encryption#101]
Currently, this project uses the YubiKey Personalization Tool which is [no longer under active development](https://www.yubico.com/support/download/yubikey-personalization-tools/). It would be nice for `yubikey-full-disk-encryption` to instead use [YubiKey Manager](https://www.yubico.com/support/download/yubikey-manager/) which has more features. I would be especially interested in this because the YubiKey Personalization Tool does not seem to support NFC, which I need.

Would you be willing to merge a pull request that implements this? I'm currently a bit busy, but I might be able to get this done in a few weeks.

**Comment by agherzan** [comment]
Yes, that is something we can/should do. `ykman` provides the top calculate/chalresp that could be used similarly. The only issue is that `calculate` takes a hex argument, so we will need to do the conversion (`od` or something). I would also vote for maintaining backward compatibility especially since everything works as expected with `yubikey-personalization` as of now.

**Comment by EliasHolzmann** [comment]
> The only issue is that calculate takes a hex argument, so we will need to do the conversion (od or something).

I'm probably being a bit obtuse right now, but didn't `ykchalresp` require the challenge in hex as well, and we get it in hex from `sha256sum` anyway? Replacing `printf %s "$YKFDE_CHALLENGE" | ykchalresp -"$YKFDE_CHALLENGE_SLOT" -i-` with `ykman otp calculate 2 "$(printf %s "$YKFDE_CHALLENGE")"` seems to work correctly for me at least (though I couldn't verify that both results are the same as I don't own a security token with USB).

> I would also vote for maintaining backward compatibility especially since everything works as expected with yubikey-personalization as of now.

Agreed, that would be nice. Although that begs the question: Which of the two yubikey packages should be a dependency, and which should be an optional dependency? Or should we specify both as optdepends?

**Comment by Vincent43** [comment]
> I'm probably being a bit obtuse right now, but didn't ykchalresp require the challenge in hex as well, and we get it in hex from sha256sum anyway?

No, ykchalresp can take anything up to 64 byte length, it doesn't need to be hex. I used sha256 only because this way it's always a valid challenge while user typed one longer than 64 byte won't be valid. 

In my testing the responses for the same challenge are different though which makes this a no-go as it breaks all current users:

```
❯ ykman otp calculate 2 9834876dcfb05cb167a5c24953eba58c4ac89b1adf57f28f2f9d09af107ee8f0
Touch your YubiKey...
63d961a7a6bfbe80f2a1e4d770a0fad1954b1d3f
❯ ykchalresp -2 9834876dcfb05cb167a5c24953eba58c4ac89b1adf57f28f2f9d09af107ee8f0
b5cbe621ae7b5c4704b55251abe3d268a0f96157
```

**Comment by Vincent43** [comment]
BTW: there is https://github.com/Frederick888/ykchalresp-nfc which we support to some [degree](https://github.com/agherzan/yubikey-full-disk-encryption#enable-nfc-support-in-ykfde-initramfs-hook-experimental).

**Comment by agherzan** [comment]
That output looks strange. What could produce that difference?

**Comment by Vincent43** [comment]
I found the fact ykman encodes challenge as hex is the answer. If I call ykchalresp in _hex mode_ then the output matches:

```
❯ ykchalresp -2 -x 61e748be34d998d8c5a13db63c2b50c4f129d2828a46bae182e6ddec5ff1ff8a
a47aa802aedcf28c6f0391d74028da887f006e8d
```

**Comment by agherzan** [comment]
I agree - and thanks for looking into it. We would need to make it backwards compatible.

**Comment by Vincent43** [comment]
Note that in Arch ykman (yubikey-manager) package depends on yubikey-personalization so all users always need them both installed.

**Comment by EliasHolzmann** [comment]
> I don't see why that dependency is defined as such. These specific tools don't have a dependency per se.

I've looked into that – it was needed until a while ago, now the dependency is unused and will probably be removed sooner or later: https://bugs.archlinux.org/task/73290

**Comment by NgoHuy** [comment]
My patch keeps old challenge working and therefor, it can replace yk-personalization.

**Comment by CapSel** [comment]
Currently `yubikey-personalization` does not detect my Yubikey. `ykman` works fine, also `yubico-authenticator-bin` from aur works fine.

**Comment by Vincent43** [comment]
which Yubikey it is? Is it over nfc?

**Comment by CapSel** [comment]
It's `Yubico.com Yubikey 4/5 U2F` as reported by `lsusb`. It has the ability to do NFC but my computer doesn't.

**Comment by Vincent43** [comment]
This doesn't tell much beside it suggests it's U2F only key that can't be used  with ykfde. Use `ykman info` and post the output with _Serial_ number removed.

Did you try to configure the key with instructions from https://github.com/agherzan/yubikey-full-disk-encryption?tab=readme-ov-file#configure-hmac-sha1-challenge-response-slot-in-yubikey ?

**Comment by CapSel** [comment]
```
Device type: Security Key NFC
Firmware version: 5.4.3
Form factor: Keychain (USB-A)
Enabled USB interfaces: FIDO
NFC transport is enabled

Applications    USB             NFC
Yubico OTP      Not available   Not available
FIDO U2F        Enabled         Enabled
FIDO2           Enabled         Enabled
OATH            Not available   Not available
PIV             Not available   Not available
OpenPGP         Not available   Not available
YubiHSM Auth    Not available   Not available
```

yubico authenticator shows that I have `FIDO2` and `FIDO U2F`. I already have 3 passkeys (max 25) on it.

command `ykpersonalize -v -3 -ochal-resp -ochal-hmac -ohmac-lt64 -oserial-api-visible -ochal-btn-trig` returns only this: `Yubikey core error: no yubikey present`.

**Comment by Vincent43** [comment]
Unfortunately you have _Yubico Security key_ which doesn't support OTP mode that is necessary for challenge-response used by ykfde. You would need _Yubikey 5_ (or 4), see https://www.yubico.com/se/store/compare/

Perhaps this is why `yubikey-personalization` doesn't detect it.

**Comment by CapSel** [comment]
Thank you for diagnosing the problem. I'll probably by 5C NFC one.

**Comment by Vincent43** [comment]
I think your key still can be used for LUKS by using fido2 approach which is supported by systemd and maybe some other tools:

https://wiki.archlinux.org/title/Systemd-cryptenroll#FIDO2_tokens

https://0pointer.net/blog/unlocking-luks2-volumes-with-tpm2-fido2-pkcs11-security-hardware-on-systemd-248.html

**Comment by CapSel** [comment]
I'll try them. Thank you very much.
